### DISCUSSION TITLE
OneOf Input Objects in Pothos?

### DISCUSSION THREAD

**Discussion by bernhardw** [hayes/pothos#673] [Q&A] [answered]
How can I use `@oneOf` Input Objects in Pothos?

Example: 
```GraphQL
input PetInput @oneOf {
  cat: CatInput
  dog: DogInput
  fish: FishInput
}

input CatInput { name: String!, numberOfLives: Int }
input DogInput { name: String!, wagsTail: Boolean }
input FishInput { name: String!, bodyLengthInMm: Int }

type Mutation {
  addPet(pet: PetInput!): Pet
}
```

Details: https://github.com/graphql/graphql-spec/pull/825

**Comment by hayes** [comment, marked as answer]
This is not yet a standardized feature, and has not landed in graphql-js, see https://github.com/graphql/graphql-js/pull/3513

This will need to be supported in the underlying graphql implementation before it lands in Pothos

**Reply by bernhardw** [reply]
Thanks a lot, makes perfect sense.

**Comment by i-tengfei** [comment]
@hayes  Any update?

**Reply by hayes** [reply]
I hadn't seen that this got back ported to GraphQL 16, I'll look into getting something added now that it's available in the graphql package

**Reply by hayes** [reply]
just pushed up some type changes that make the isOneOf option define stricter types for builder.inputObject.  Thats a type-only change, the functionality for the oneOf directive was already fully supported

**Reply by vecerek** [reply]
Are there any plans to generate a tagged union type from `oneOf` input objects? Any way I can help?

**Reply by hayes** [reply]
Can you elaborate on how you would expect that to work, and what the issue is with the current implementation?

**Reply by vecerek** [reply]
Of course 🙂 There is not necessarily an issue with the current implementation. In my opinion, [tagged unions](https://en.wikipedia.org/wiki/Tagged_union), or sum types, represent the concept of `oneOf` more correctly in the type system compared to the currently generated types. Let me demonstrate this through an example.

Given the following `oneOf` input schema:
```ts
export const OneOfExample = builder.inputType(
  "OneOfExampleInput",
  {
    isOneOf: true,
    fields: (t) => ({
      a: t.int(),
      b: t.string(),
      c: t.boolean(),
      d: t.id(),
    })
  }
);
```

the generated TS type looks like this:

```ts
type OneOfExample =
  | { a: number; b?: undefined; c?: undefined; d?: undefined; }
  | { b: string; a?: undefined; c?: undefined; d?: undefined; }
  | { c: boolean; a?: undefined; b?: undefined; d?: undefined; }
  | { d: string; a?: undefined; b?: undefined; c?: undefined; }
```

This checks all the functional requirements but notice how the list of undefined keys grows with the number of fields in the `oneOf` input type. However, in practice, only one of the fields can be present at any one time, so why should we represent all the undefined options for each case? The developer experience also suffers because the IDE still presents all options:

<img width="1017" alt="Screenshot 2024-10-19 at 14 24 25" src="https://github.com/user-attachments/assets/8f38f6bb-d05c-4008-a40b-8f24b3f36c54">

A much better representation of `oneOf` is a tagged union which gets rid of the "absurd" options from the type signature. It would look something like this instead:

```ts
type OneOfExample =
  | { _tag: "a"; a: number; }
  | { _tag: "b"; b: string; }
  | { _tag: "c"; c: boolean; }
  | { _tag: "d"; d: string; }
```

Not only we get rid of all the noise from the type signature but the developer experience becomes also more favorable:

<img width="727" alt="Screenshot 2024-10-19 at 14 25 38" src="https://github.com/user-attachments/assets/b72f3b4f-11b1-4712-8924-725e48a43889">

<img width="1021" alt="Screenshot 2024-10-19 at 14 25 57" src="https://github.com/user-attachments/assets/9f89bc83-0401-49be-9aa9-5e48e218f25d">

It is a solution that scales better with the number of fields in the input type.

**Reply by hayes** [reply]
The issue with this is that it requires a runtime change which kinda goes against Pothos design principles.  Pothos pretty strictly helps you build schemas that don't contain any Pothos specific login at runtime.  For the most part everything that executes to resolve a query (when not using plugins) comes either from the graphql package, or user defined resolvers.

To make this work, we'd have to wrap any resolver of fields that accept oneOf inputs, and transform the input to add the tag.  This isn't trivial when the inputs are nested deeply in other input types.

If the graphql package added a tag, I would definitely  update the types to reflect that, but adding runtime overhead for type convenience isn't really a good fit for core behavior.  

This would be easy to do as a plugin though if you wanted to implement this yourself for your own schema
